pkg/generate/code/synced.go

```diff
@@ -0,0 +1,204 @@
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package code
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws-controllers-k8s/code-generator/pkg/fieldpath"
+	ackgenconfig "github.com/aws-controllers-k8s/code-generator/pkg/generate/config"
+	"github.com/aws-controllers-k8s/code-generator/pkg/model"
+
+	awssdkmodel "github.com/aws/aws-sdk-go/private/model/api"
+)
+
+// ResourceIsSynced returns the Go code that verifies whether a resource is synced or
+// not. This code is generated using ack-generate configuration.
+// See ack-generate/pkg/config.SyncedConfiguration.
+//
+//  Sample output:
+//
+//  	candidates0 := []string{"AVAILABLE", "ACTIVE"}
+//  	if !ackutil.InStrings(*r.ko.Status.TableStatus, candidates0) {
+//  		return false, nil
+//  	}
+//  	if r.ko.Spec.ProvisionedThroughput == nil {
+//  		return false, nil
+//  	}
+//  	if r.ko.Spec.ProvisionedThroughput.ReadCapacityUnits == nil {
+//  		return false, nil
+//  	}
+//  	candidates1 := []int{0, 10}
+//  	if !ackutil.InStrings(*r.ko.Spec.ProvisionedThroughput.ReadCapacityUnits, candidates1) {
+//  		return false, nil
+//  	}
+//  	candidates2 := []int{0}
+//  	if !ackutil.InStrings(*r.ko.Status.ItemCount, candidates2) {
+//  		return false, nil
+//  	}
+func ResourceIsSynced(
+	cfg *ackgenconfig.Config,
+	r *model.CRD,
+	// String
+	resVarName string,
+	// Number of levels of indentation to use
+	indentLevel int,
+) string {
+	out := "\n"
+	resConfig, ok := cfg.ResourceConfig(r.Names.Original)
+	if !ok || resConfig.Synced == nil || len(resConfig.Synced.When) == 0 {
+		return out
+	}
+
+	for _, condition := range resConfig.Synced.When {
+		if condition.Path == nil || *condition.Path == "" {
+			panic("Received an empty sync condition path. 'SyncCondition.Path' must be provided.")
+		}
+		if len(condition.In) == 0 {
+			panic("'SyncCondition.In' must be provided.")
+		}
+		fp := fieldpath.FromString(*condition.Path)
+		field, err := getTopLevelField(r, *condition.Path)
+		if err != nil {
+			msg := fmt.Sprintf("cannot find top level field of path '%s': %v", *condition.Path, err)
+			panic(msg)
+		}
+		candidatesVarName := fmt.Sprintf("%sCandidates", field.Names.CamelLower)
+		if fp.Size() == 2 {
+			out += scalarFieldEqual(resVarName, candidatesVarName, field.ShapeRef.GoTypeElem(), condition)
+		} else {
+			out += fieldPathSafeEqual(resVarName, candidatesVarName, field, condition)
+		}
+	}
+
+	return out
+}
+
+func getTopLevelField(r *model.CRD, fieldPath string) (*model.Field, error) {
+	fp := fieldpath.FromString(fieldPath)
+	if fp.Size() < 2 {
+		return nil, fmt.Errorf("fieldPath must contain at least two elements, received: %s", fieldPath)
+	}
+
+	head := fp.PopFront()
+	fieldName := fp.PopFront()
+	switch head {
+	case "Spec":
```

Comment:
Note that this hard-coding makes Crossplane unable to use any of this... because their top-level fields are under `Spec.ForProvider` which is specified in the [`PrefixConfig`](https://github.com/aws-controllers-k8s/code-generator/blob/6da05cd784c388f3b113778d034960398c2dd56e/pkg/generate/config/config.go#L34) of the generator configuration. This is one of those places we need to be aware of when refactoring the code generator. /cc @brycahta